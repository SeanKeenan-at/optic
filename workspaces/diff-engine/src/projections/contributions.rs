use crate::RfcEvent;
use cqrs_core::{Aggregate, AggregateEvent, Event};
use std::collections::HashMap;

#[derive(Debug, Clone)]
pub struct ContributionsProjection {
  // TODO check whether this is best ds to use
  pub hash_map: HashMap<String, HashMap<String, String>>,
}

impl ContributionsProjection {
  pub fn to_json_string(&self) -> String {
    serde_json::to_string(&self.hash_map).expect("hash map should be serializable")
  }

  pub fn with_contribution(&mut self, id: String, contribution_key: String, value: String) {
    self
      .hash_map
      .entry(id)
      .or_insert_with(|| HashMap::new())
      .insert(contribution_key, value);
  }
}

impl Default for ContributionsProjection {
  fn default() -> Self {
    ContributionsProjection {
      hash_map: HashMap::new(),
    }
  }
}

impl Aggregate for ContributionsProjection {
  fn aggregate_type() -> &'static str {
    "contributions_projection"
  }
}

impl<I> From<I> for ContributionsProjection
where
  I: IntoIterator,
  I::Item: AggregateEvent<Self>,
{
  fn from(events: I) -> Self {
    let mut projection = ContributionsProjection::default();
    for event in events.into_iter() {
      projection.apply(event);
    }
    projection
  }
}

// Events
impl AggregateEvent<ContributionsProjection> for RfcEvent {
  fn apply_to(self, projection: &mut ContributionsProjection) {
    match self {
      RfcEvent::ContributionAdded(e) => projection.with_contribution(e.id, e.key, e.value),
      _ => eprintln!(
        "Ignoring applying event of type '{}' for '{}'",
        self.event_type(),
        ContributionsProjection::aggregate_type()
      ),
    }
  }
}
