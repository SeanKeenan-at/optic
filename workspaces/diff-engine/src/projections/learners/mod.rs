pub mod undocumented_bodies;
