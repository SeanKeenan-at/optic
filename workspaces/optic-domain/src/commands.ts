/*
TODO Replace with Rust constructors eventually
*/

// TODO fix any types in this file
export type CQRSCommand =
  | AddShapeType
  | AddQueryParametersType
  | AddRequestType
  | FieldFromShapeType
  | AddFieldType
  | ShapeProviderType
  | ProviderInShapeType
  | SetParameterShapeType
  | SetFieldShapeType
  | AddShapeParameterType
  | RemoveFieldType
  | AddPathParameterType
  | AddPathComponentType
  | AddContributionType
  | ShapedBodyDescriptorType
  | SetQueryParameteresShapeType
  | SetResponseBodyShapeType
  | SetRequestBodyShapeType
  | AddResponseByPathAndMethodType
  | PrunePathComponentsType
  | QueryParametersShapeDescriptor;

export type AddShapeType = ReturnType<typeof AddShape>;
export type AddQueryParametersType = ReturnType<typeof AddQueryParameters>;
export type AddRequestType = ReturnType<typeof AddRequest>;
export type FieldFromShapeType = ReturnType<typeof FieldShapeFromShape>;
export type AddFieldType = ReturnType<typeof AddField>;
export type ShapeProviderType = ReturnType<typeof ShapeProvider>;
export type ProviderInShapeType = ReturnType<typeof ProviderInShape>;
export type SetParameterShapeType = ReturnType<typeof SetParameterShape>;
export type SetFieldShapeType = ReturnType<typeof SetFieldShape>;
export type AddShapeParameterType = ReturnType<typeof AddShapeParameter>;
export type RemoveFieldType = ReturnType<typeof RemoveField>;
export type AddPathParameterType = ReturnType<typeof AddPathParameter>;
export type AddPathComponentType = ReturnType<typeof AddPathComponent>;
export type AddContributionType = ReturnType<typeof AddContribution>;
export type ShapedBodyDescriptorType = ReturnType<typeof ShapedBodyDescriptor>;
export type SetQueryParameteresShapeType = ReturnType<
  typeof SetQueryParametersShape
>;
export type QueryParametersShapeDescriptor = ReturnType<
  typeof QueryParametersShapeDescriptor
>;
export type SetResponseBodyShapeType = ReturnType<typeof SetResponseBodyShape>;
export type SetRequestBodyShapeType = ReturnType<typeof SetRequestBodyShape>;
export type AddResponseByPathAndMethodType = ReturnType<
  typeof AddResponseByPathAndMethod
>;
export type PrunePathComponentsType = ReturnType<typeof PrunePathComponents>;

export function AddShape(
  shapeId: string,
  baseShapeId: string,
  name: string = ''
) {
  return { AddShape: { shapeId, baseShapeId, name } };
}

export function AddQueryParameters(
  httpMethod: string,
  pathId: string,
  queryParametersId: string
) {
  return { AddQueryParameters: { httpMethod, pathId, queryParametersId } };
}

export function AddRequest(
  httpMethod: string,
  pathId: string,
  requestId: string
) {
  return { AddRequest: { httpMethod, pathId, requestId } };
}

export function FieldShapeFromShape(fieldId: string, shapeId: string) {
  return { FieldShapeFromShape: { fieldId, shapeId } };
}

export function AddField(
  fieldId: string,
  shapeId: string,
  name: string,
  shapeDescriptor: FieldFromShapeType
) {
  return { AddField: { fieldId, shapeId, name, shapeDescriptor } };
}

export function ShapeProvider(shapeId: string) {
  return { ShapeProvider: { shapeId } };
}

export function ProviderInShape(
  shapeId: string,
  providerDescriptor: any,
  consumingParameterId: string
) {
  return {
    ProviderInShape: { shapeId, providerDescriptor, consumingParameterId },
  };
}

export function SetParameterShape(shapeDescriptor: any) {
  return { SetParameterShape: { shapeDescriptor } };
}
export function SetFieldShape(shapeDescriptor: FieldFromShapeType) {
  return { SetFieldShape: { shapeDescriptor } };
}

export function AddShapeParameter(
  shapeParameterId: string,
  shapeId: string,
  name: string
) {
  return { AddShapeParameter: { shapeParameterId, shapeId, name } };
}

export function RemoveField(fieldId: string) {
  return { RemoveField: { fieldId } };
}

export function AddPathParameter(
  pathId: string,
  parentPathId: string,
  name: string
) {
  return {
    AddPathParameter: {
      pathId,
      parentPathId,
      name,
    },
  };
}

export function AddPathComponent(
  pathId: string,
  parentPathId: string,
  name: string
) {
  return {
    AddPathComponent: {
      pathId,
      parentPathId,
      name,
    },
  };
}

export function AddContribution(id: string, key: string, value: string) {
  return {
    AddContribution: {
      id,
      key,
      value,
    },
  };
}

export function ShapedBodyDescriptor(
  httpContentType: string,
  shapeId: string,
  isRemoved: boolean = false
) {
  return {
    httpContentType,
    shapeId,
    isRemoved,
  };
}

export function QueryParametersShapeDescriptor(
  shapeId: string,
  isRemoved: boolean = false
) {
  return {
    shapeId,
    isRemoved,
  };
}

export function SetQueryParametersShape(
  queryParametersId: string,
  shapeDescriptor: ReturnType<typeof QueryParametersShapeDescriptor>
) {
  return {
    SetQueryParametersShape: {
      queryParametersId,
      shapeDescriptor,
    },
  };
}

export function SetResponseBodyShape(
  responseId: string,
  bodyDescriptor: ReturnType<typeof ShapedBodyDescriptor>
) {
  return {
    SetResponseBodyShape: {
      responseId,
      bodyDescriptor,
    },
  };
}
export function SetRequestBodyShape(
  requestId: string,
  bodyDescriptor: ReturnType<typeof ShapedBodyDescriptor>
) {
  return {
    SetRequestBodyShape: {
      requestId,
      bodyDescriptor,
    },
  };
}

export function AddResponseByPathAndMethod(
  responseId: string,
  pathId: string,
  httpMethod: string,
  httpStatusCode: number
) {
  return {
    AddResponseByPathAndMethod: {
      responseId,
      pathId,
      httpMethod,
      httpStatusCode,
    },
  };
}

export function PrunePathComponents() {
  return {
    PrunePathComponents: {},
  };
}
