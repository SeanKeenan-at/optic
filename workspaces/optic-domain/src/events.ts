export enum ICoreShapeKinds {
  ObjectKind = '$object',
  ListKind = '$list',
  MapKind = '$map',
  OneOfKind = '$oneOf',
  AnyKind = '$any',
  StringKind = '$string',
  NumberKind = '$number',
  BooleanKind = '$boolean',
  NullableKind = '$nullable',
  OptionalKind = '$optional',
  UnknownKind = '$unknown',
}
export enum ICoreShapeInnerParameterNames {
  ListInner = '$listItem',
  NullableInner = '$nullableInner',
  OptionalInner = '$optionalInner',
}
