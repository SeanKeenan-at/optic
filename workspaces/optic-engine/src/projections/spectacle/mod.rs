pub mod endpoints;
pub mod shape;
