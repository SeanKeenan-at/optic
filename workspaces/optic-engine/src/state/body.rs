use crate::shapehash;
use serde::de::value;
use serde_json::map::Map as JsonMap;
use serde_json::Value as JsonValue;
use serde_urlencoded;
use std::collections::{BTreeMap, HashMap};

#[derive(PartialEq, Clone, Debug, Hash, Eq)]
pub enum BodyDescriptor {
  Object(ObjectDescriptor),
  Array(ItemsDescriptor),
  String,
  Number,
  Boolean,
  Null,
}

impl BodyDescriptor {
  pub fn empty_object() -> Self {
    Self::Object(ObjectDescriptor::from(std::iter::empty()))
  }
}

#[derive(PartialEq, Clone, Debug, Hash, Eq)]
pub struct FieldDescriptor(pub String, pub Box<BodyDescriptor>);

impl FieldDescriptor {
  pub fn new(key: String, body: BodyDescriptor) -> Self {
    Self(key, Box::new(body))
  }
}

#[derive(PartialEq, Clone, Debug, Hash, Eq)]
pub struct ObjectDescriptor {
  fields: Vec<FieldDescriptor>,
}

impl ObjectDescriptor {
  pub fn keys(&self) -> impl Iterator<Item = &String> {
    self.fields.iter().map(|FieldDescriptor(key, body)| key)
  }

  pub fn entries(self) -> impl Iterator<Item = (String, BodyDescriptor)> {
    self
      .fields
      .into_iter()
      .map(|FieldDescriptor(key, body)| (key, *body))
  }
}

impl<T> From<T> for ObjectDescriptor
where
  T: Iterator<Item = (String, BodyDescriptor)>,
{
  fn from(entries: T) -> Self {
    Self {
      fields: entries
        .map(|(key, value)| FieldDescriptor::new(key, value))
        .collect(),
    }
  }
}

#[derive(PartialEq, Clone, Debug, Hash, Eq)]
pub struct ItemsDescriptor {
  unique_items: Box<Vec<(BodyDescriptor, Vec<usize>)>>,
}

impl<T> From<T> for ItemsDescriptor
where
  T: Iterator<Item = BodyDescriptor>,
{
  fn from(all_items: T) -> Self {
    let mut unique_items = all_items
      .enumerate()
      .fold(HashMap::new(), |mut indexes_by_unique_items, (i, item)| {
        {
          let items = indexes_by_unique_items.entry(item).or_insert(vec![]);
          items.push(i);
        };

        indexes_by_unique_items
      })
      .into_iter()
      .map(|(item, indexes)| (item.clone(), indexes))
      .collect::<Vec<_>>();

    unique_items.sort_by_key(|(_, indexes)| *indexes.first().unwrap());

    Self {
      unique_items: Box::new(unique_items),
    }
  }
}

impl From<BodyDescriptor> for ItemsDescriptor {
  fn from(single_item: BodyDescriptor) -> Self {
    Self {
      unique_items: Box::new(vec![(single_item, vec![0])]),
    }
  }
}

impl ItemsDescriptor {
  // @TODO: decide if this is still interesting at all. Could be reconstructed from unique_items
  // pub fn into_all(self) -> impl Iterator<Item = BodyDescriptor> {
  //
  // }

  pub fn into_unique(self) -> impl Iterator<Item = (BodyDescriptor, Vec<usize>)> {
    self.unique_items.into_iter()
  }

  pub fn unique_items_count(&self) -> usize {
    self.unique_items.len()
  }
}

impl From<shapehash::ShapeDescriptor> for BodyDescriptor {
  fn from(mut shape_hash_descriptor: shapehash::ShapeDescriptor) -> Self {
    match shape_hash_descriptor.field_type {
      shapehash::ShapeDescriptor_PrimitiveType::OBJECT => {
        let fields = shape_hash_descriptor
          .take_fields()
          .into_iter()
          .map(|field_descriptor| {
            (
              field_descriptor.key.clone(),
              BodyDescriptor::from(field_descriptor.hash.unwrap()),
            )
          });
        BodyDescriptor::Object(ObjectDescriptor::from(fields))
      }
      shapehash::ShapeDescriptor_PrimitiveType::ARRAY => {
        let items = shape_hash_descriptor
          .take_items()
          .into_iter()
          .map(|item_descriptor| BodyDescriptor::from(item_descriptor));

        BodyDescriptor::Array(ItemsDescriptor::from(items))
      }
      shapehash::ShapeDescriptor_PrimitiveType::BOOLEAN => BodyDescriptor::Boolean,
      shapehash::ShapeDescriptor_PrimitiveType::NULL => BodyDescriptor::Null,
      shapehash::ShapeDescriptor_PrimitiveType::NUMBER => BodyDescriptor::Number,
      shapehash::ShapeDescriptor_PrimitiveType::STRING => BodyDescriptor::String,
    }
  }
}

impl From<JsonValue> for BodyDescriptor {
  fn from(json_value: JsonValue) -> Self {
    match json_value {
      JsonValue::Object(json_fields) => {
        let fields = json_fields
          .into_iter()
          .map(|(key, value)| (key, BodyDescriptor::from(value)));

        BodyDescriptor::Object(ObjectDescriptor::from(fields))
      }
      JsonValue::Array(json_items) => {
        let items = json_items
          .into_iter()
          .map(|item_descriptor| BodyDescriptor::from(item_descriptor));

        BodyDescriptor::Array(ItemsDescriptor::from(items))
      }
      JsonValue::Bool(_) => BodyDescriptor::Boolean,
      JsonValue::Null => BodyDescriptor::Null,
      JsonValue::Number(_) => BodyDescriptor::Number,
      JsonValue::String(_) => BodyDescriptor::String,
    }
  }
}

impl From<String> for BodyDescriptor {
  fn from(string: String) -> Self {
    BodyDescriptor::String
  }
}

impl From<&String> for BodyDescriptor {
  fn from(str: &String) -> Self {
    BodyDescriptor::String
  }
}

#[derive(Debug, Default)]
pub struct ParsedQueryString {
  entries: Vec<(String, String)>,
}

impl ParsedQueryString {
  pub fn from_str(query_string: &str) -> Result<Self, serde_urlencoded::de::Error> {
    let entries = serde_urlencoded::from_str(query_string)?;
    Ok(Self { entries })
  }
}

impl From<ParsedQueryString> for BodyDescriptor {
  fn from(parsed_qs: ParsedQueryString) -> Self {
    let mut values_by_key = BTreeMap::new();

    for (key, value) in parsed_qs.entries {
      let entry = values_by_key.entry(key).or_insert_with(|| vec![]);
      entry.push(value);
    }

    let fields = values_by_key.into_iter().map(|(key, values)| {
      let value_descriptor = if values.len() == 1 {
        BodyDescriptor::String
      } else {
        let item_descriptor = ItemsDescriptor::from(BodyDescriptor::String); // we only ever expect strings with this parser
        BodyDescriptor::Array(item_descriptor)
      };

      (key, value_descriptor)
    });

    BodyDescriptor::Object(ObjectDescriptor::from(fields))
  }
}

#[cfg(test)]
mod test {
  use super::*;
  use insta::assert_debug_snapshot;

  #[test]
  fn query_string_can_be_parsed_to_body_descriptor() {
    let parsed = ParsedQueryString::from_str("foo=bar&csv=1,2,51&list=twelve&list=fourteen")
      .expect("should be able to parse a query string");

    assert_debug_snapshot!(
      "query_string_can_be_parsed_to_body_descriptor__parsed",
      &parsed
    );

    let body_descriptor = BodyDescriptor::from(parsed);

    assert_debug_snapshot!(
      "query_string_can_be_parsed_to_body_descriptor__body_descriptor",
      body_descriptor
    );
  }
}
