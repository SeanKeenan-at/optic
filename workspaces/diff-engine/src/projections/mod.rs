pub mod conflicts;
pub mod endpoint;
pub mod history;
pub mod shape;
pub mod spec_events;

pub use conflicts::ConflictsProjection;
pub use endpoint::EndpointProjection;
pub use history::{CommitId, HistoryProjection};
pub use shape::ShapeProjection;
pub use spec_events::{SpecAssemblerError, SpecAssemblerProjection};

use crate::events::{EndpointEvent, RfcEvent, ShapeEvent, SpecEvent};
use cqrs_core::{Aggregate, AggregateCommand, AggregateEvent, CommandError};
use std::error::Error;

#[derive(Debug)]
pub struct SpecProjection {
  endpoint: endpoint::EndpointProjection,
  history: history::HistoryProjection,
  shape: shape::ShapeProjection,
  conflicts: conflicts::ConflictsProjection,
}

impl Default for SpecProjection {
  fn default() -> Self {
    Self {
      endpoint: EndpointProjection::default(),
      history: HistoryProjection::default(),
      shape: ShapeProjection::default(),
      conflicts: ConflictsProjection::default(),
    }
  }
}

impl SpecProjection {
  pub fn endpoint(&self) -> &EndpointProjection {
    &self.endpoint
  }

  pub fn history(&self) -> &HistoryProjection {
    &self.history
  }

  pub fn shape(&self) -> &ShapeProjection {
    &self.shape
  }
  pub fn conflicts(&self) -> &ConflictsProjection {
    &self.conflicts
  }
}

impl Aggregate for SpecProjection {
  fn aggregate_type() -> &'static str {
    "spec_projection"
  }
}

// Events
// ------

impl AggregateEvent<SpecProjection> for SpecEvent {
  fn apply_to(self, projection: &mut SpecProjection) {
    match self {
      SpecEvent::EndpointEvent(event) => {
        projection.endpoint.apply(event.clone());
        projection.conflicts.apply(event);
      }
      SpecEvent::ShapeEvent(event) => {
        projection.shape.apply(event.clone());
        projection.conflicts.apply(event);
      }
      SpecEvent::RfcEvent(event) => projection.history.apply(event),
    }
  }
}

impl AggregateEvent<HistoryProjection> for SpecEvent {
  fn apply_to(self, projection: &mut HistoryProjection) {
    if let SpecEvent::RfcEvent(event) = self {
      event.apply_to(projection);
    }
  }
}

impl<I> From<I> for SpecProjection
where
  I: IntoIterator,
  I::Item: AggregateEvent<Self>,
{
  fn from(events: I) -> Self {
    let mut projection = SpecProjection::default();
    for event in events.into_iter() {
      projection.apply(event);
    }
    projection
  }
}

// Convient application of spec event variants
// -------------------------------------------
impl AggregateEvent<SpecProjection> for ShapeEvent {
  fn apply_to(self, projection: &mut SpecProjection) {
    projection.apply(SpecEvent::ShapeEvent(self))
  }
}

impl AggregateEvent<SpecProjection> for RfcEvent {
  fn apply_to(self, projection: &mut SpecProjection) {
    projection.apply(SpecEvent::RfcEvent(self))
  }
}

impl AggregateEvent<SpecProjection> for EndpointEvent {
  fn apply_to(self, projection: &mut SpecProjection) {
    projection.apply(SpecEvent::EndpointEvent(self))
  }
}
