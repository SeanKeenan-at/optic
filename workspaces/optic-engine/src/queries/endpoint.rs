use crate::commands::{EndpointCommand, SpecCommand};
use crate::events::HttpInteraction;
use crate::projections::endpoint::{Edge, EndpointProjection, Node, ROOT_PATH_ID};
use crate::projections::endpoint::{RequestBodyDescriptor, ResponseBodyDescriptor};
use crate::state::endpoint::{
  HttpMethod, HttpStatusCode, PathComponentId, PathComponentIdRef, RequestId, ResponseId,
};
use petgraph::graph::Graph;
use petgraph::visit::EdgeFilteredNeighborsDirected;
use serde::Serialize;

pub struct EndpointQueries<'a> {
  pub endpoint_projection: &'a EndpointProjection,
}

impl<'a> EndpointQueries<'a> {
  pub fn new(endpoint_projection: &'a EndpointProjection) -> Self {
    EndpointQueries {
      endpoint_projection,
    }
  }

  fn extract_normalized_path(path: &str) -> &str {
    if path.eq("/") {
      path
    } else if path.ends_with("/") {
      &path[..path.len() - 1]
    } else {
      path
    }
  }

  pub fn resolve_interaction_path(
    &self,
    interaction: &HttpInteraction,
  ) -> Option<PathComponentIdRef> {
    self.resolve_path(&interaction.request.path)
  }

  pub fn resolve_path(&self, path: &str) -> Option<PathComponentIdRef> {
    let path = Self::extract_normalized_path(path);
    // eprintln!("{}", path);
    let mut path_components = path.split('/');
    // skip leading empty
    path_components.next();
    let mut last_resolved_path_id = Some(ROOT_PATH_ID);
    while let Some(s) = path_components.next() {
      // eprintln!("trying to match segment {}", s);
      let node_index = self
        .graph_get_index(last_resolved_path_id.unwrap())
        .expect("expected a node with node_id to exist");

      last_resolved_path_id = None;

      // first search for named path components
      let children = self.graph_get_children(node_index);
      for child in children {
        let child_node = self.endpoint_projection.graph.node_weight(child).unwrap();

        match child_node {
          Node::PathComponent(id, descriptor) => {
            // eprintln!("1 - neighbor {}", descriptor.name);
            if !descriptor.is_parameter {
              if descriptor.name == s {
                last_resolved_path_id = Some(id);
                break;
              }
            }
          }
          _ => {}
        }
      }
      if let Some(x) = last_resolved_path_id {
        continue;
      }
      // try path parameters since we didn't find any matches in regular path components
      let children = self.graph_get_children(node_index);
      for child in children {
        let child_node = self.endpoint_projection.graph.node_weight(child).unwrap();
        match child_node {
          Node::PathComponent(id, descriptor) => {
            // eprintln!("2 - neighbor {}", descriptor.name);

            if descriptor.is_parameter {
              last_resolved_path_id = Some(id);
              break;
            }
          }
          _ => {}
        }
      }
      if let Some(x) = last_resolved_path_id {
        continue;
      }
      return None;
    }
    last_resolved_path_id
  }

  pub fn resolve_operations_by_request_method(
    &self,
    method: &'a String,
    path_id: PathComponentIdRef,
  ) -> impl Iterator<Item = (&RequestId, &RequestBodyDescriptor)> {
    self
      .resolve_requests(path_id, method)
      .expect("expected a operations to exist")
  }

  pub fn resolve_operations(
    &self,
    interaction: &'a HttpInteraction,
    path_id: PathComponentIdRef,
  ) -> impl Iterator<Item = (&RequestId, &RequestBodyDescriptor)> {
    self.resolve_operations_by_request_method(&interaction.request.method, path_id)
  }

  pub fn resolve_requests(
    &self,
    path_id: PathComponentIdRef,
    method: &'a String,
  ) -> Option<impl Iterator<Item = (&RequestId, &RequestBodyDescriptor)>> {
    let path_node_index = self.graph_get_index(path_id)?;
    let children = self
      .endpoint_projection
      .graph
      .neighbors_directed(*path_node_index, petgraph::Direction::Incoming);
    let matching_method = children
      .filter(move |i| {
        let node = self.endpoint_projection.graph.node_weight(*i).unwrap();
        match node {
          Node::HttpMethod(http_method) => method == http_method,
          _ => false,
        }
      })
      .flat_map(move |i| {
        let children = self
          .endpoint_projection
          .graph
          .neighbors_directed(i, petgraph::Direction::Incoming);

        let operations = children.filter_map(move |i| {
          let node = self.endpoint_projection.graph.node_weight(i).unwrap();
          match node {
            Node::Request(request_id, request_descriptor) => Some((request_id, request_descriptor)),
            _ => None,
          }
        });
        operations
      });
    Some(matching_method)
  }

  pub fn resolve_request_by_method_and_content_type(
    &self,
    path_id: PathComponentIdRef,
    method: &'a String,
    content_type: Option<&'a String>,
  ) -> Option<(&RequestId, &RequestBodyDescriptor)> {
    self.resolve_requests(path_id, method).and_then(|mut it| {
      it.find(|(id, body)| match content_type {
        Some(content_type) => match body.body {
          Some(ref body) => body.http_content_type.eq(content_type),
          None => false,
        },
        None => body.body.is_none(),
      })
    })
  }

  pub fn resolve_responses(
    &self,
    path_id: &'a PathComponentId,
    method: &'a String,
  ) -> Option<impl Iterator<Item = (&ResponseId, &ResponseBodyDescriptor)>> {
    let response_nodes = self
      .endpoint_projection
      .get_response_nodes(path_id, method)?;

    Some(response_nodes.map(|node| match node {
      Node::Response(response_id, body_descriptor) => (response_id, body_descriptor),
      _ => unreachable!("get response nodes should only return response nodes"),
    }))
  }

  pub fn resolve_response_by_method_status_code_and_content_type(
    &self,
    path_id: PathComponentIdRef,
    method: &'a str,
    status_code: u16,
    content_type: Option<&'a String>,
  ) -> Option<(&ResponseId, &ResponseBodyDescriptor)> {
    self
      .resolve_responses_by_method_and_status_code(method, status_code, path_id)
      .find(|(id, body)| match content_type {
        Some(content_type) => match body.body {
          Some(ref body) => body.http_content_type.eq(content_type),
          None => false,
        },
        None => body.body.is_none(),
      })
  }

  pub fn resolve_responses_by_method_and_status_code(
    &self,
    method: &'a str,
    status_code: u16,
    path_id: PathComponentIdRef,
  ) -> impl Iterator<Item = (&ResponseId, &ResponseBodyDescriptor)> {
    let path_node_index = self
      .graph_get_index(path_id)
      .expect("expected a node with node_id to exist");
    let children = self
      .endpoint_projection
      .graph
      .neighbors_directed(*path_node_index, petgraph::Direction::Incoming);
    let matching_status_code = children
      .filter(move |i| {
        let node = self.endpoint_projection.graph.node_weight(*i).unwrap();
        // eprintln!("method node {:?}", node);
        match node {
          Node::HttpMethod(http_method) => method == *http_method,
          _ => false,
        }
      })
      .flat_map(move |i| {
        // eprintln!("method node id {:?}", i);
        let children = self
          .endpoint_projection
          .graph
          .neighbors_directed(i, petgraph::Direction::Incoming);
        let status_code_nodes = children.filter_map(move |i| {
          // eprintln!("method child node id {:?}", i);
          let node = self.endpoint_projection.graph.node_weight(i).unwrap();
          match node {
            Node::HttpStatusCode(http_status_code) => {
              // eprintln!("status code {:?}", http_status_code);
              if status_code == *http_status_code {
                Some(i)
              } else {
                None
              }
            }
            _ => None,
          }
        });
        status_code_nodes
      })
      .flat_map(move |i| {
        let children = self
          .endpoint_projection
          .graph
          .neighbors_directed(i, petgraph::Direction::Incoming);
        let response_nodes = children.filter_map(move |i| {
          // eprintln!("status_code child node id {:?}", i);
          let node = self.endpoint_projection.graph.node_weight(i).unwrap();
          // eprintln!("status_code child node {:?}", node);
          match node {
            Node::Response(response_id, response_descriptor) => {
              Some((response_id, response_descriptor))
            }
            _ => None,
          }
        });
        response_nodes
      });
    matching_status_code
  }

  pub fn delete_endpoint_commands(
    &self,
    path_id: &'a PathComponentId,
    method: &'a HttpMethod,
  ) -> Option<DeleteEndpointCommands> {
    let request_ids = self
      .resolve_requests(path_id, method)?
      .map(|(request_id, _)| request_id);
    let response_ids = self
      .resolve_responses(path_id, method)?
      .map(|(response_id, _)| response_id);

    let request_commands = request_ids.cloned().map(EndpointCommand::remove_request);
    let response_commands = response_ids.cloned().map(EndpointCommand::remove_response);

    Some(DeleteEndpointCommands {
      path_id: path_id.clone(),
      method: method.clone(),
      commands: request_commands
        .chain(response_commands)
        .map(SpecCommand::from)
        .collect(),
    })
  }

  fn graph_get_index(&self, node_id: &str) -> Option<&petgraph::graph::NodeIndex> {
    self.endpoint_projection.node_id_to_index.get(node_id)
  }

  fn graph_get_node(&self, node_id: &str) -> &Node {
    self
      .endpoint_projection
      .graph
      .node_weight(
        *self
          .graph_get_index(node_id)
          .expect("expected a node with node_id to exist"),
      )
      .expect("expected node with node_id to exist")
  }

  fn graph_get_children(
    &self,
    node_index: &petgraph::graph::NodeIndex,
  ) -> petgraph::graph::Neighbors<Edge> {
    let neighbors = self
      .endpoint_projection
      .graph
      .neighbors_directed(*node_index, petgraph::Direction::Incoming);
    return neighbors;
  }
}

#[derive(Debug, Serialize)]
pub struct DeleteEndpointCommands {
  path_id: PathComponentId,
  method: HttpMethod,
  commands: Vec<SpecCommand>,
}

#[cfg(test)]
mod test {
  use super::*;
  use crate::events::SpecEvent;
  use crate::projections::SpecProjection;
  use crate::Aggregate;
  use petgraph::dot::Dot;
  use serde_json::json;

  fn interaction_with_path(path: String) -> HttpInteraction {
    serde_json::from_value(json!(
      {
        "uuid": "id",
        "request": {
          "host": "example.com",
          "method": "GET",
          "path": path,
          "query": {
            "shapeHashV1Base64": null,
            "asJsonString": null,
            "asText": null
          },
          "headers": {
            "shapeHashV1Base64": null,
            "asJsonString": null,
            "asText": null
          },
          "body": {
            "contentType": null,
            "value": {
              "shapeHashV1Base64": null,
              "asJsonString": null,
              "asText": null
            }
          }
        },
        "response": {
          "statusCode": 200,
          "headers": {
            "shapeHashV1Base64": null,
            "asJsonString": null,
            "asText": null
          },
          "body": {
            "contentType": null,
            "value": {
              "shapeHashV1Base64": null,
              "asJsonString": null,
              "asText": null
            }
          }
        },
        "tags": []
      }
    ))
    .unwrap()
  }
  #[test]
  pub fn can_ignore_trailing_slash() {
    let interaction: HttpInteraction = interaction_with_path(String::from("/a/b/c/"));
    let normalized_path = EndpointQueries::extract_normalized_path(&interaction.request.path);
    assert_eq!(normalized_path, "/a/b/c")
  }

  #[test]
  pub fn can_handle_no_trailing_slash() {
    let interaction: HttpInteraction = interaction_with_path(String::from("/a/b/c"));
    let normalized_path = EndpointQueries::extract_normalized_path(&interaction.request.path);
    assert_eq!(normalized_path, "/a/b/c")
  }
  #[test]
  pub fn can_handle_root_path() {
    let interaction: HttpInteraction = interaction_with_path(String::from("/"));
    let normalized_path = EndpointQueries::extract_normalized_path(&interaction.request.path);
    assert_eq!(normalized_path, "/")
  }

  #[test]
  pub fn can_generate_delete_commands() {
    let events: Vec<SpecEvent> = serde_json::from_value(json!([
      {"PathComponentAdded": { "pathId": "path_1", "parentPathId": "root", "name": "posts" }},
      {"PathComponentAdded": { "pathId": "path_2", "parentPathId": "path_1", "name": "favourites" }},
      {"PathComponentAdded": { "pathId": "path_3", "parentPathId": "root", "name": "authors" }},

      {"RequestAdded": { "requestId": "request_1", "pathId": "path_2", "httpMethod": "GET"}},
      {"ResponseAddedByPathAndMethod": {"responseId": "response_1", "pathId": "path_2", "httpMethod": "GET", "httpStatusCode": 200 }},

      {"RequestAdded": { "requestId": "request_2", "pathId": "path_2", "httpMethod": "POST"}},
      {"ResponseAddedByPathAndMethod": {"responseId": "response_2", "pathId": "path_2", "httpMethod": "POST", "httpStatusCode": 201 }},

      {"RequestAdded": { "requestId": "request_3", "pathId": "path_3", "httpMethod": "GET"}},
      {"ResponseAddedByPathAndMethod": {"responseId": "response_3", "pathId": "path_3", "httpMethod": "GET", "httpStatusCode": 200 }},
    ]))
    .expect("should be able to deserialize test events");

    let spec_projection = SpecProjection::from(events);
    // dbg!(Dot::with_config(&spec_projection.endpoint().graph, &[]));

    let endpoint_queries = EndpointQueries::new(spec_projection.endpoint());

    let subject_path = String::from("path_2");
    let subject_method = String::from("GET");
    let deleted_endpoint_commmands = endpoint_queries
      .delete_endpoint_commands(&subject_path, &subject_method)
      .expect("delete commands are generated for existing path and method");

    let updated_spec =
      assert_valid_commands(spec_projection.clone(), deleted_endpoint_commmands.commands);
    let updated_queries = EndpointQueries::new(&updated_spec.endpoint());
    let remaining_requests = updated_queries
      .resolve_requests(&subject_path, &subject_method)
      .unwrap()
      .collect::<Vec<_>>();
    let remaining_responses = updated_queries
      .resolve_responses(&subject_path, &subject_method)
      .unwrap()
      .collect::<Vec<_>>();

    // TODO: enable these assertions as the EndpointProjection handles the resulting events
    // assert_eq!(remaining_requests.len(), 0);
    // assert_eq!(remaining_responses.len(), 0);
  }

  fn assert_valid_commands(
    mut spec_projection: SpecProjection,
    commands: impl IntoIterator<Item = SpecCommand>,
  ) -> SpecProjection {
    // let mut spec_projection = SpecProjection::default();
    for command in commands {
      let events = spec_projection
        .execute(command)
        .expect("generated commands must be valid");

      for event in events {
        spec_projection.apply(event)
      }
    }

    spec_projection
  }
}
