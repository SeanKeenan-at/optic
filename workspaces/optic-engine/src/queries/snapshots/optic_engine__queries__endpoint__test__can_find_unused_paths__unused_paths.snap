---
source: workspaces/optic-engine/src/queries/endpoint.rs
expression: unused_path_ids
---
[
    "path_3",
    "path_parameter_1",
]
